FannyAndersson/PayWay
# Service Worker & Push Notifications

# Hello!

This PR introduces:

- Service worker
- Push notifications to multiple devices

There is one minor issue I am aware of, and it is the following:

- When upgrading version of service worker, old push notification subscriptions for the user in the database do not get removed, but new ones get added. It is not sustainable, but at this point, it doesn't really matter.

There are probably more issues lurking about, so be vigilant in your testing.

## How to test:

- npm install in root directory to get new module

- create a file called vapid-keys.json in root directory, info is in slack

## What to test:

- [ ] Offline functionality
- [ ] Push notifications - do you get them when you expect to, and not get them when you don't expect to?
- [ ] Caching strategy

The caching strategy is the following:

In development, everything is network only (nothing gets cached)
In production, everything apart from:

- Chrome extensions
- Web sockets
- Calls to anything that starts with /api/

gets cached and served as cache-first - the above requests are served as network only, so the app is not really usable offline.

It is **very important** to manually update version number in service worker when code changes - otherwise it will serve old stuff. 

### How to test production:

0. Make sure you take down the dev servers so they are not running.
1. Build frontend for production by executing `npm run build` in **frontend directory**.
2. Run post build script by executing `node postBuildScript` in **frontend directory**
3. Start backend in production mode with `npm run backend` in **root directory**
4. Visit http://localhost:3001/ in your browser of choice
